### ISSUE TITLE
Color Tone Problem on SDL2/macOS when not using sRGB profile

### ISSUE THREAD

**Issue by palxex** [sdlpal/sdlpal#77]
From [PCem emulation of Win95 version](https://tieba.baidu.com/p/5548308155), I found the color tone is notably difference from the sdlpal SDL2 version. Then I made more tests. Finally I found that there're two groups( no matter DOS/Win95 resource used ): one is PCem/dosbox, the other is palx/sdlpal/win95 perfect patch; the previous group seems more like the scene in my memories. Consider all the latter 3 uses the rgb*4 VGA palette conversation method, is it means there're a more precious method?
![2018-02-17 11 56 51](https://user-images.githubusercontent.com/58222/36343046-c73c6b24-1441-11e8-803e-a784dee65d11.png)

**Comment by zephray** [comment]
The problem is actually how to map a 18bit RGB666 color used by VGA to a 24bit RGB888 color. Consider only one channel, 6bit  means 0-63, if it gets multiplied by 4 (or shift left by 2bits), the new range is 0-252, close enough to 0-255 of a 8bit number. But if precision is really important, a better method exists, which is new = (old << 2) | (old >> 6). This method will map 0-63 to 0-255 with full scale linearly.

**Comment by zephray** [comment]
However I am not sure if this is the cause of the difference of color tone. I think the difference between these two methods should be very subtle and unnoticeable if its just 6bit to 8bit. (But will give a noticeable difference when doing 16bit RGB565 to 24bit RGB888 conversion, and a huge difference when doing 4bit 16 shades of gray to 24bit RGB888 conversion)

**Comment by louyihua** [comment]
@palxex @zephray 
From code's perspective, the mutliply by 4 approach is the way used by PAL Win95 itself, as the following code shows:
```assembly
PAL!intpate:
053b13d6 55              push    ebp
053b13d7 8bec            mov     ebp,esp
053b13d9 83ec04          sub     esp,4
053b13dc 53              push    ebx
053b13dd 56              push    esi
053b13de 57              push    edi
053b13df 833d30c03b0500  cmp     dword ptr [PAL!cvlong+0x7eb7 (053bc030)],0
053b13e6 0f840d000000    je      PAL!intpate+0x23 (053b13f9)
053b13ec 833d34c03b0500  cmp     dword ptr [PAL!cvlong+0x7ebb (053bc034)],0
053b13f3 0f8505000000    jne     PAL!intpate+0x28 (053b13fe)
053b13f9 e9f6000000      jmp     PAL!intpate+0x11e (053b14f4)
053b13fe c7055cc03b0500000000 mov dword ptr [PAL!cvlong+0x7ee3 (053bc05c)],0
053b1408 c70504153c0501000000 mov dword ptr [PAL!cvlong+0xd38b (053c1504)],1
053b1412 c745fc00000000  mov     dword ptr [ebp-4],0
053b1419 e903000000      jmp     PAL!intpate+0x4b (053b1421)
053b141e ff45fc          inc     dword ptr [ebp-4]
053b1421 817dfc00010000  cmp     dword ptr [ebp-4],100h
053b1428 0f8d8a000000    jge     PAL!intpate+0xe2 (053b14b8)
053b142e 8b45fc          mov     eax,dword ptr [ebp-4]
053b1431 8d0440          lea     eax,[eax+eax*2]
053b1434 8b4d08          mov     ecx,dword ptr [ebp+8]
053b1437 33d2            xor     edx,edx
053b1439 8a1408          mov     dl,byte ptr [eax+ecx]
053b143c 8d049500000000  lea     eax,[edx*4]
053b1443 8b4dfc          mov     ecx,dword ptr [ebp-4]
053b1446 88048d90063c05  mov     byte ptr PAL!cvlong+0xc517 (053c0690)[ecx*4],al
053b144d 8b45fc          mov     eax,dword ptr [ebp-4]
053b1450 8d0440          lea     eax,[eax+eax*2]
053b1453 8b4d08          mov     ecx,dword ptr [ebp+8]
053b1456 33d2            xor     edx,edx
053b1458 8a540801        mov     dl,byte ptr [eax+ecx+1]
053b145c 8d049500000000  lea     eax,[edx*4]
053b1463 8b4dfc          mov     ecx,dword ptr [ebp-4]
053b1466 88048d91063c05  mov     byte ptr PAL!cvlong+0xc518 (053c0691)[ecx*4],al
053b146d 8b45fc          mov     eax,dword ptr [ebp-4]
053b1470 8d0440          lea     eax,[eax+eax*2]
053b1473 8b4d08          mov     ecx,dword ptr [ebp+8]
053b1476 33d2            xor     edx,edx
053b1478 8a540802        mov     dl,byte ptr [eax+ecx+2]
053b147c 8d049500000000  lea     eax,[edx*4]
053b1483 8b4dfc          mov     ecx,dword ptr [ebp-4]
053b1486 88048d92063c05  mov     byte ptr PAL!cvlong+0xc519 (053c0692)[ecx*4],al
053b148d 8b45fc          mov     eax,dword ptr [ebp-4]
053b1490 c6048593063c0500 mov     byte ptr PAL!cvlong+0xc51a (053c0693)[eax*4],0
053b1498 8b45fc          mov     eax,dword ptr [ebp-4]
053b149b 833c8590063c0500 cmp     dword ptr PAL!cvlong+0xc517 (053c0690)[eax*4],0
053b14a3 0f840a000000    je      PAL!intpate+0xdd (053b14b3)
053b14a9 c70504153c0500000000 mov dword ptr [PAL!cvlong+0xd38b (053c1504)],0
053b14b3 e966ffffff      jmp     PAL!intpate+0x48 (053b141e)
053b14b8 6a00            push    0
053b14ba 6a01            push    1
053b14bc a130c03b05      mov     eax,dword ptr [PAL!cvlong+0x7eb7 (053bc030)]
053b14c1 50              push    eax
053b14c2 a130c03b05      mov     eax,dword ptr [PAL!cvlong+0x7eb7 (053bc030)]
053b14c7 8b00            mov     eax,dword ptr [eax]
053b14c9 ff5058          call    dword ptr [eax+58h]
053b14cc 6890063c05      push    offset PAL!cvlong+0xc517 (053c0690)
053b14d1 6800010000      push    100h
053b14d6 6a00            push    0
053b14d8 6a00            push    0
053b14da a140c03b05      mov     eax,dword ptr [PAL!cvlong+0x7ec7 (053bc040)]
053b14df 50              push    eax
053b14e0 a140c03b05      mov     eax,dword ptr [PAL!cvlong+0x7ec7 (053bc040)]
053b14e5 8b00            mov     eax,dword ptr [eax]
053b14e7 ff5018          call    dword ptr [eax+18h]
053b14ea c7055cc03b0501000000 mov dword ptr [PAL!cvlong+0x7ee3 (053bc05c)],1
053b14f4 5f              pop     edi
053b14f5 5e              pop     esi
053b14f6 5b              pop     ebx
053b14f7 c9              leave
053b14f8 c20400          ret     4
```
In the above code, `[ebp-4]` is the looping counter, `[ebp+8]` is the input 3-byte, 6-bit palette, and `PAL!cvlong+0xc517` is the starting point of global palette buffer which will be used by DirectDraw's API. This code use `lea` for calculating, so this code snippet:
```assembly
053b142e 8b45fc          mov     eax,dword ptr [ebp-4]
053b1431 8d0440          lea     eax,[eax+eax*2]
053b1434 8b4d08          mov     ecx,dword ptr [ebp+8]
053b1437 33d2            xor     edx,edx
053b1439 8a1408          mov     dl,byte ptr [eax+ecx]
053b143c 8d049500000000  lea     eax,[edx*4]
053b1443 8b4dfc          mov     ecx,dword ptr [ebp-4]
053b1446 88048d90063c05  mov     byte ptr PAL!cvlong+0xc517 (053c0690)[ecx*4],al
```
It multiplys looping counter by 3 (`lea     eax,[eax+eax*2]`), gets the byte from the input palette (`mov     dl,byte ptr [eax+ecx]`), multiplys this byte by 4 (`lea     eax,[edx*4]`) and writes it back the global palette buffer (`mov     byte ptr PAL!cvlong+0xc517 (053c0690)[ecx*4],al`).

So I don't think the way palx/sdlpal/patch deals with palette is wrong. This difference may be caused by postprocessing approaches.

**Comment by palxex** [comment]
Yestoday all tests performs on macOS (patch tested on wine). Just tested again on windows, the difference of dosbox/sdlpal/patch seems disappears... I'll test on FreeDOS later.

**Comment by palxex** [comment]
Finally it turns out a Color Profile setting problem in my Mac- - sorry disturbing.
The way I discover it is by suggestion of @PalMusicFan, I compiled a SDL1 version, and find in this version, the color tone is same as DOSBox - so the problem is surely has no relation with palette color conversation. Then I tested again on a optimus laptop, and ensured all color is same on iGPU and dGPU. Then the only questionable place is my device; after digging I found the problem.
Previous, I uses ICC profile delivered from my display device vendor AOC. In this profile, SDL2( the shader version ) will renders all color a little cold than original, which differ from SDL1, the pure 2D version. After I switched to the built-in sRGB profile, the two version now produces same color - and surely, which also confirmed same as windows versions.
Some more questions: 
1. why when using vendor-supplied ICC profile, it does not looks like original version?
2. why in sRGB and only in sRGB profile the render result is same? Is it has any relationship with tone mapping?

**Comment by palxex** [comment]
Got answers:
For 2, because sRGB is a common minimized color gamut.
For 1, seems the current RGB colors that SDL2 built-in shaders outputs has overcome sRGB.
For 3, so we just need to limit all output color inside sRGB.
So the resolution method is : for OpenGL renderer backend, add this line
```
SDL_GL_SetAttribute(SDL_GL_FRAMEBUFFER_SRGB_CAPABLE, 1);
```
will solves everything. Not sure what need to do for other backend - maybe not needed at all.

**Comment by palxex** [comment]
Some facts:
1. My display device need a ICC profile to display sRGB correctly; from the profile, seems the device's built-in gamut is slightly larger than sRGB. That's why on my device, when using sRGB profile, the pixels seems a little dark, not as windows machines.
3. OpenGL does not have a color gamut definition, in SDL2 case, we'd assume it uses sRGB gamut. That's why on my display, when using sRGB profile, openGL results same as 2D pixel - which both darker than normal.
3. According to [this technote](https://developer.apple.com/library/content/technotes/tn2313/_index.html#//apple_ref/doc/uid/DTS40014694-CH1-NONCOLORMANAGEDFRAMEWORKS-OPENGL___EXPLICIT_COLOR_MANAGEMENT_EXAMPLE), on macOS, openGL is not color-managed, which means ColorSync will not modify its output color. That's why on my device, when switching to any other profile - even P3, the openGL output still seems same, which means same as sRGB result - darker too.
4. Apple gives us two choice: 1 do nothing. 2 manually do color conversation, as the mentioned demo ColorWhirled. Since the latter one has been deprecated so long - the demo it self only available in xcode 1.0, which programs for asm glsl - and even the API it uses has been marked as legacy/deprecated too; and the new ColorSync API that replaces it still poor in documentation, and seems only available at CPU side, not to say on iOS it's a private framework.
5. Since this problem only addressable on my device, and may have performance impact, seems I'd better let it there.

Some resource may useful for others who have interests:
1. [ColorWhirled.zip](https://github.com/sdlpal/sdlpal/files/1747843/ColorWhirled.zip) from xcode 1.0
2. [test.zip](https://github.com/sdlpal/sdlpal/files/1747847/test.zip) a minimized openGL texture rendering demo; if you see no difference with what it displays and any image viewer, then on your display settings you'll never find difference as the first image shows.

Closing the issue, thanks for discussing:)

**Comment by palxex** [comment]
Some tests on macOS after SDL 2.0.8 official release:
1. on OpenGL backend, the problem still as same as before; without objective-c code modification( https://github.com/sdlpal/SDL/commit/a0c986556295661e76c10700aea4b41ea3032a84, which cannot be done in SDLPal code due to c/objc runloop interop problem), color calibration still not works, which means when you switching color profile in Display Preferences, nothing will happen in SDLPal; it also means on my monitor, the color displays incorrect, since color calibration was a must for it.
2. on new introduced Metal backend, since its not really ready for industrial usage(for example, texture locking not implemented), I'd modify both SDLPal and SDL2( includes objective-c and metal shader) to make it able to run. After that, the result is the color calibration is already applied. 

These results leads to another strange point: Apple points(in above document/WWDC video mentioned ) that both OpenGL and Metal are default "Color UNMANAGED" , for performance consideration, which could be manually turned on by specify colorspace, as what I did in the mentioned SDL 2.0.5 patch. But in SDL 2.0.8, I did not saw any such code snippet.
